Add OpenVINO Installation and GPU Availability Test Cases

frenchwr opened the pull request:
This uses the same flow as before for getting the environment all set up:

```
$ snapcraft
$ sudo snap install --dangerous --classic ./checkbox-dss_2.0_amd64.snap
$ checkbox-dss.install-dss-deps
$ checkbox-dss.install-full-deps # for docker
$ checkbox-dss.gpu-plugin-validation
$ checkbox-dss.remove-dss-deps
```

For the GPU availability test, I took [the code from this notebook](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/108-gpu-device/108-gpu-device.ipynb) in the Intel OpenVINO Toolkit GitHub repo.

frenchwr committed:
Add openvino test cases

mckees commented:
Can you please add the docker dependencies into the install-dss-deps bin so that we can get rid of install-full-deps?

frenchwr commented:
Also I just updated the PR description to include a link to the code I used for checking GPU availability.

mckees commented on checkbox-provider-dss/units/jobs.pxu:
All Jobs should start with `set -e`. This is not the only instance, but I'll just mention it here

frenchwr replied:
Good catch, I definitely need it here. I was thinking there may be some simple if-else tests where it's better to handle the error in some way. For example, when we're doing something like `result=$(microk8s.kubectl get service...)` we could check for an empty result and error out with something like "Failure: no K8s service found". But probably it's still informative if the command fails to actually stop and show stderr, so I guess it wouldn't hurt to include `set -e` in these cases as well. Don't mind me I'm just arguing with myself. Nothing to see here. :)

frenchwr replied:
done

mckees commented on checkbox-provider-dss/units/jobs.pxu:
should have an extra "_" before the {{

frenchwr replied:
fixed.

mckees commented on checkbox-provider-dss/units/jobs.pxu:
Mine returns 
```
Test failure: expected result to contain app name 'notebook-openvino' but got 
```

So it looks like this comes back blank. Presumably this is because of the Docker failure

frenchwr replied:
Good catch - done.

mckees commented on checkbox-provider-dss/units/jobs.pxu:
This job fails for me. Docker does not start
```
Building OpenVINO Docker image locally. Please be patient.
ERROR: permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Head "http://%2Fvar%2Frun%2Fdocker.sock/_ping": dial unix /var/run/docker.sock: connect: permission denied
-------------------------------
```
This seems to make all tests after fail

frenchwr replied:
Yeah I think we're missing post-install steps: https://docs.docker.com/engine/install/linux-postinstall/

I'm going to reprovision `mustang` (once MAAS is back up) and test from a clean slate.

frenchwr replied:
Added the recommended postinstall scripts and confirmed these steps work on a newly provisioned system with a iGPU

frenchwr committed:
Move docker install to install-dss-deps script

frenchwr committed:
Ensure all tests exit on error

frenchwr commented:
Ok - I think I got everything. Feel free to re-test.

mckees commented:
Thanks for the new changes. I'm still seeing these failures on a clean system:
```
-----------------------------[ Running job 3 / 10 ]-----------------------------
--------------------[ Check number of available daemonsets ]--------------------
ID: com.canonical.certification::intel_gpu_plugin/daemonset_check_number_available_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test failure: expected numberAvailable to be 1 but got 
--------------------------------------------------------------------------------
Outcome: job failed
-----------------------------[ Running job 4 / 10 ]-----------------------------
----------------------[ Check number of ready daemonsets ]----------------------
ID: com.canonical.certification::intel_gpu_plugin/daemonset_check_number_ready_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test failure: expected numberReady to be 1 but got 0
--------------------------------------------------------------------------------
Outcome: job failed
-----------------------------[ Running job 7 / 10 ]-----------------------------
---------------------[ Launch OpenVINO+Jupyter container ]----------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_start_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Building OpenVINO Docker image locally. Please be patient.
ERROR: failed to solve: failed to read dockerfile: failed to fetch remote https://github.com/openvinotoolkit/openvino_notebooks: git error: exit status 128
stderr:
error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
error: 600 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output

--------------------------------------------------------------------------------
Outcome: job failed
-----------------------------[ Running job 8 / 10 ]-----------------------------
-------------------------[ Check OpenVINO Deployment ]--------------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_deployment_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test failure: expected result to contain app name 'notebook-openvino' but got 
--------------------------------------------------------------------------------
Outcome: job failed
-----------------------------[ Running job 9 / 10 ]-----------------------------
------[ Check jupyter and openvino python packages in OpenVINO container ]------
ID: com.canonical.certification::intel_gpu_plugin/openvino_libs_installed_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Outcome: job failed
----------------------------[ Running job 10 / 10 ]-----------------------------
----------------------[ Check OpenVINO GPU Availability ]-----------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_gpu_avail_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Outcome: job failed
``

mckees commented:
It looks like #3 and #4 succeeded on a second run through with seemingly irrelevant cleanup changes, so it's possible that a) we have intermittent failures on these tests b) we have some initialization happening in the later stages of the testing which allows a second run through to pass or c) my cleanup changes do something I'm not aware of

frenchwr committed:
Add retries for daemonset health tests

frenchwr commented:
Ok - I've added a brief delay into the test that deploys the Intel K8s GPU Device Plugin, and also added retries to the tests for the daemonset health. TIL that `kubectl rollout status` can return before all the status fields (e.g. `numberAvailable`, `numberReady` for the daemonset have actually updated). I'm definitely learning as I go. :) 

Everything on a freshly provisioned machine (`mustang` in the Canonical Locker) passed for me [1]. I'll paste the exact steps [2] I used below in case you want to try to reproduce.

As for the other failure you saw when attempting to docker build, I think you may be hitting a resource limit on your machine...maybe network or storage related? Where are you running the tests?

[1]

```shell
$ checkbox-dss.gpu-plugin-validation 
WARNING:root:remote is deprecated and will be removed in the next major release of Checkbox. Please use control instead
WARNING:plainbox.providers.v1:$PROVIDERPATH is defined, so following provider sources are ignored ['/snap/checkbox-dss/x1/providers/checkbox-provider-dss', '/home/ubuntu/.local/share/plainbox-providers-1', '/var/tmp/checkbox-providers-develop'] 
==========[ Bootstrap com.canonical.certification::executable (1/3) ]===========
=============[ Bootstrap com.canonical.certification::snap (2/3) ]==============
=========[ Bootstrap com.canonical.certification::graphics_card (3/3) ]=========
-----------------------------[ Running job 1 / 10 ]-----------------------------
--------------------[ Install Intel K8s GPU Device Plugin ]---------------------
ID: com.canonical.certification::intel_gpu_plugin/microk8s_install_intel_gpu_plugin_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
namespace/node-feature-discovery created
customresourcedefinition.apiextensions.k8s.io/nodefeaturerules.nfd.k8s-sigs.io created
customresourcedefinition.apiextensions.k8s.io/nodefeatures.nfd.k8s-sigs.io created
serviceaccount/nfd-master created
serviceaccount/nfd-worker created
role.rbac.authorization.k8s.io/nfd-worker created
clusterrole.rbac.authorization.k8s.io/nfd-master created
rolebinding.rbac.authorization.k8s.io/nfd-worker created
clusterrolebinding.rbac.authorization.k8s.io/nfd-master created
configmap/nfd-master-conf created
configmap/nfd-worker-conf created
service/nfd-master created
deployment.apps/nfd-master created
daemonset.apps/nfd-worker created
nodefeaturerule.nfd.k8s-sigs.io/intel-dp-devices created
nodefeaturerule.nfd.k8s-sigs.io/intel-gpu-platform-labeling created
daemonset.apps/intel-gpu-plugin created
Waiting for daemon set spec update to be observed...
Waiting for daemon set "nfd-worker" rollout to finish: 0 out of 1 new pods have been updated...
Waiting for daemon set "nfd-worker" rollout to finish: 0 of 1 updated pods are available...
daemon set "nfd-worker" successfully rolled out
daemon set "intel-gpu-plugin" successfully rolled out
Test success: Intel K8s GPU Device Plugin deployed.
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 2 / 10 ]-----------------------------
----------------------------[ Check DaemonSet Name ]----------------------------
ID: com.canonical.certification::intel_gpu_plugin/daemonset_check_name_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test success: 'intel-gpu-plugin' daemonset is deployed!
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 3 / 10 ]-----------------------------
--------------------[ Check number of available daemonsets ]--------------------
ID: com.canonical.certification::intel_gpu_plugin/daemonset_check_number_available_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Info: received empty numberAvailable result. Sleeping for 10 seconds. Retry 1/10.
Test success: 1 daemonset in numberAvailable status.
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 4 / 10 ]-----------------------------
----------------------[ Check number of ready daemonsets ]----------------------
ID: com.canonical.certification::intel_gpu_plugin/daemonset_check_number_ready_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test success: 1 daemonset in numberReady status.
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 5 / 10 ]-----------------------------
------------[ Initialize DSS environment with common k8s manifests ]------------
ID: com.canonical.certification::intel_gpu_plugin/dss_initialize_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
namespace/dss created
persistentvolumeclaim/mlflow created
deployment.apps/mlflow-deployment created
service/mlflow created
persistentvolumeclaim/notebooks-home created
deployment.apps/mlflow-deployment condition met
Test success: common DSS manifests deployed.
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 6 / 10 ]-----------------------------
-----------------------------[ Check Service Name ]-----------------------------
ID: com.canonical.certification::intel_gpu_plugin/dss_check_service_name_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test success: 'mlflow' service is deployed!
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 7 / 10 ]-----------------------------
---------------------[ Launch OpenVINO+Jupyter container ]----------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_start_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Building OpenVINO Docker image locally. Please be patient.
sha256:a7a2d0d0cba19366deba12b26e3de4874c3823a104f7b0353f2803f10f163536
sha256:e1e4c6e26ad6e086b687a062a5c498b35b2466311253500432840a5514fccbda
unpacking docker.io/local/openvinotoolkit:render-addon (sha256:304eb808224a9ff980c75a6d31eb0ab0896f2f28f1228ecd4e2c96cab7699f48)...done
deployment.apps/notebook-openvino created
service/user-notebook-openvino created
deployment.apps/notebook-openvino condition met
Test success: OpenVINO+Jupyter container is running.
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 8 / 10 ]-----------------------------
-------------------------[ Check OpenVINO Deployment ]--------------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_deployment_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Test success: 'notebook-openvino' app is deployed!
--------------------------------------------------------------------------------
Outcome: job passed
-----------------------------[ Running job 9 / 10 ]-----------------------------
------[ Check jupyter and openvino python packages in OpenVINO container ]------
ID: com.canonical.certification::intel_gpu_plugin/openvino_libs_installed_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Examining pod notebook-openvino-58599754fb-264qh...
Name: jupyter
Version: 1.0.0
Summary: Jupyter metapackage. Install all the Jupyter components in one go.
Home-page: http://jupyter.org
Author: Jupyter Development Team
Author-email: jupyter@googlegroups.org
License: BSD
Location: /opt/app-root/lib/python3.8/site-packages
Requires: ipykernel, ipywidgets, jupyter-console, nbconvert, notebook, qtconsole
Required-by: treon
Name: openvino
Version: 2024.2.0
Summary: OpenVINO(TM) Runtime
Home-page: https://docs.openvino.ai/2023.0/index.html
Author: Intel(R) Corporation
Author-email: openvino@intel.com
License: OSI Approved :: Apache Software License
Location: /opt/app-root/lib/python3.8/site-packages
Requires: numpy, openvino-telemetry, packaging
Required-by: openvino-dev
Test success: jupyter and openvino python packages installed in container.
--------------------------------------------------------------------------------
Outcome: job passed
----------------------------[ Running job 10 / 10 ]-----------------------------
----------------------[ Check OpenVINO GPU Availability ]-----------------------
ID: com.canonical.certification::intel_gpu_plugin/openvino_gpu_avail_i915
Category: OpenCL regression tests
--------------------------------------------------------------------------------
Examining pod notebook-openvino-58599754fb-264qh...
Test success: Intel GPU device (Intel(R) Graphics [0xa780] (iGPU)) available to OpenVINO.
--------------------------------------------------------------------------------
Outcome: job passed
==================================[ Results ]===================================
  job passed   : Enumerate available system executables
  job passed   : Install Intel K8s GPU Device Plugin
  job passed   : Check DaemonSet Name
  job passed   : Check number of available daemonsets
  job passed   : Check number of ready daemonsets
  job passed   : Initialize DSS environment with common k8s manifests
  job passed   : Check Service Name
  job passed   : Launch OpenVINO+Jupyter container
  job passed   : Check OpenVINO Deployment
  job passed   : Check jupyter and openvino python packages in OpenVINO container
  job passed   : Check OpenVINO GPU Availability
file:///home/ubuntu/.local/share/checkbox-ng/submission_2024-06-18T18.19.02.806019.html
file:///home/ubuntu/.local/share/checkbox-ng/submission_2024-06-18T18.19.02.806019.junit.xml
file:///home/ubuntu/.local/share/checkbox-ng/submission_2024-06-18T18.19.02.806019.tar.xz
```

[2]

* Provision from MAAS UI with 22.04 + HWE kernel

```shell
# From laptop (this assumes you're testing on mustang)
$ scp -r checkbox-dss-validation/ mustang:~ # from laptop
# Remaining steps performed on remote server
$ sudo snap install snapcraft --classic
$ sudo snap install checkbox22
$ lxd init --auto
$ cd checkbox-dss-validation
$ snapcraft
$ sudo snap install --dangerous --classic ./checkbox-dss_2.0_amd64.snap
$ checkbox-dss.install-dss-deps
$ checkbox-dss.gpu-plugin-validation
$ checkbox-dss.remove-dss-deps
```

mckees commented:
Alright, all checks pass. LGTM
